File: ruby/aa/pair_sum.rb
require 'set'

def pair_sum(nums, k)
  pairs = Set.new
  seen = Set.new

  nums.each do |num|
    target = k - num
    if seen.include?(target)
      pairs << [[num, target].min, [num, target].max]
    end
    seen << num
  end

  pairs
end

p pair_sum([1, 2, -1], 0)          # => #<Set: {[-1, 1]}>
p pair_sum([1, 2, -1, -1], 0)      # => #<Set: {[-1, 1]}>
p pair_sum([1, 2, -1, -1, -2], 0)  # => #<Set: {[-1, 1], [-2, 2]}>
p pair_sum([1, 2, -1, -1, -2], 1)  # => #<Set: {[-1, 2]}>
p pair_sum([1, 2, -1, -1, -2], -1) # => #<Set: {[-2, 1]}>

File: js/aa/pairSum.js
function pairSum (nums, k) {
  const pairs = {};
  const seen = {};

  for (let i = 0; i < nums.length; i++) {
    let target = k - nums[i];
    if (seen[target]) {
      let min = Math.min(nums[i], target);
      let max = Math.max(nums[i], target);
      pairs[[min, max]] = true;
    }
    seen[nums[i]] = true;
  }

  return Object.keys(pairs);
}

console.log(pairSum([1, 2, -1], 0));
console.log(pairSum([1, 2, -1, -1], 0));
console.log(pairSum([1, 2, -1, -1, -2], 0));
console.log(pairSum([1, 2, -1, -1, -2], 1));
console.log(pairSum([1, 2, -1, -1, -2], -1));
